diff --git a/src/create.luau b/src/create.luau
index db33663..35257ad 100644
--- a/src/create.luau
+++ b/src/create.luau
@@ -52,7 +52,7 @@ local function create(class_or_instance: string|Instance, properties: { [unknown
         else constructor
 end
 
-type Instances = {
+export type Instances = {
     Folder: Folder,
     BillboardGui: BillboardGui,
     CanvasGroup: CanvasGroup,
@@ -74,7 +74,7 @@ type Instances = {
     WorldModel: WorldModel,
 }
 
-type function Properties(instance: type?)
+export type function Properties(instance: type?)
     local properties = types.newtable()
 
     while instance do
diff --git a/src/init.luau b/src/init.luau
index 11d1bd2..0a76880 100644
--- a/src/init.luau
+++ b/src/init.luau
@@ -1,10 +1,12 @@
 assert(game, "when using vide outside of Roblox, require lib.luau instead")
 
-local vide = require(script.lib)
+local vide = require("@self/lib")
 
 export type source<T> = vide.source<T>
 export type Source<T> = vide.Source<T>
 export type context<T> = vide.context<T>
 export type Context<T> = vide.Context<T>
+export type Instances = vide.Instances
+export type Properties<Name> = vide.Properties<Name>
 
 return vide
diff --git a/src/lib.luau b/src/lib.luau
index 28784b4..7fc3916 100644
--- a/src/lib.luau
+++ b/src/lib.luau
@@ -27,6 +27,8 @@ export type Source<T> = source.Source<T>
 export type source<T> = Source<T>
 export type Context<T> = context.Context<T>
 export type context<T> = Context<T>
+export type Instances = create.Instances
+export type Properties<Name> = create.Properties<Name>
 
 local function step(dt: number)
     if game then debug.profilebegin("VIDE STEP") end
